File: typescript/packages/core/src/commands/cli/builtin/git/show.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult } from '../../../../io/types.ts'
import type { CommandFnResult } from '../../../config.ts'
import { FlagView } from '../../../spec/flag_view.ts'
import type { CLIInvocation } from '../../types.ts'
import { GitError } from './errors.ts'
import {
  MEDIUM,
  needsDecorations,
  oneline,
  parsePretty,
  presetBlock,
  renderTemplate,
  type CommitFacts,
  type Decorations,
  type LogFormat,
} from './format.ts'
import { decorations, prettyValue } from './history.ts'
import {
  joinOutput,
  commitOutput,
  parseDiffFlags,
  renamesEnabled,
  type DiffFlags,
} from './diff_output.ts'
import { commitFacts, configBool, opened } from './repo.ts'
import { resolveCommit } from './revparse.ts'
import { checkOperands, escaped, fatal, revisionArg } from './util.ts'
import { encodeText } from '../../../../shell/bytes.ts'

/**
 * The parsed shape of a `git show` invocation.
 *
 * `--no-ext-diff` is accepted but carries no field: there are no external
 * diff drivers here, so it changes nothing by construction.
 */
interface ShowFlags {
  readonly diff: DiffFlags
  readonly pretty: LogFormat
  readonly date: string
}

/** Read the raw show flag kwargs into a frozen struct. */
function parseShowFlags(fl: FlagView, defaultRenames = true, quotePathFully = true): ShowFlags {
  const spelled = prettyValue(fl)
  return {
    date: fl.asStr('date') ?? 'default',
    diff: parseDiffFlags(fl, true, 'dense-combined', true, defaultRenames, quotePathFully),
    pretty: spelled !== null ? parsePretty(spelled) : MEDIUM,
  }
}

/**
 * The commit header in the requested format.
 *
 * `format:` is a separator, so a single commit prints with no trailing
 * newline at all; `tformat:` terminates the entry even when it renders
 * empty, except that an empty template prints nothing, matching
 * `log --format=`. Pinned against git 2.37 and 2.54.
 */
function header(
  commit: CommitFacts,
  flags: ShowFlags,
  width: number,
  decor: Decorations | null,
): string {
  const fmt = flags.pretty
  if (fmt.kind === 'oneline') return `${oneline(commit, width)}\n`
  if (fmt.kind === 'format' || fmt.kind === 'tformat') {
    const text = renderTemplate(fmt.template ?? '', commit, width, decor, flags.date)
    if (fmt.kind === 'tformat') {
      return fmt.template === null || fmt.template === '' ? '' : `${text}\n`
    }
    return text
  }
  return `${presetBlock(commit, fmt.kind, width, flags.date).join('\n')}\n`
}

/** Show one commit: its log entry, then its diff against its parent. */
export async function show(inv: CLIInvocation): Promise<CommandFnResult> {
  const doors = inv.doors ?? {}
  const texts = [...inv.texts]
  const fl = new FlagView(inv.flags)
  try {
    checkOperands(texts, undefined, escaped(inv.argv))
    const repo = await opened(fl, doors)
    const parsed = parseShowFlags(
      fl,
      await renamesEnabled(repo),
      await configBool(repo, 'core.quotepath', true),
    )
    const oid = await resolveCommit(repo, revisionArg(texts))
    const facts = await commitFacts(repo, oid)
    const decor = needsDecorations(parsed.pretty) ? await decorations(repo) : null
    const head = header(facts, parsed, repo.abbrev, decor)
    const bodies = await commitOutput(repo, facts, parsed.diff)
    return [
      encodeText(
        joinOutput(
          facts,
          head,
          bodies,
          parsed.pretty.kind,
          repo.abbrev,
          parsed.diff.summary && !parsed.diff.noPatch,
        ),
      ),
      new IOResult(),
    ]
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
}

export async function diffTree(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  try {
    const repo = await opened(fl, inv.doors ?? {})
    const commit = await commitFacts(repo, await resolveCommit(repo, inv.texts[0] ?? 'HEAD'))
    const bodies = await commitOutput(
      repo,
      commit,
      parseDiffFlags(fl, false, 'off', false, true, await configBool(repo, 'core.quotepath', true)),
      fl.asBool('r'),
      false,
    )
    const out = bodies
      .map((body) => (fl.asBool('no_commit_id') ? '' : commit.oid + '\n') + body)
      .join('')
    return [encodeText(out), new IOResult()]
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
}

File: python/mirage/commands/cli/builtin/git/show.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
from dataclasses import dataclass

from dulwich.objects import Commit
from dulwich.repo import BaseRepo

from mirage.commands.cli.builtin.git.diff_output import (DiffFlags,
                                                         commit_output,
                                                         join_output,
                                                         parse_diff_flags,
                                                         renames_enabled)
from mirage.commands.cli.builtin.git.errors import GitError, NoWorkspaceError
from mirage.commands.cli.builtin.git.format import (MEDIUM, Decorations,
                                                    LogFormat,
                                                    needs_decorations, oneline,
                                                    parse_pretty, preset_block,
                                                    render_template)
from mirage.commands.cli.builtin.git.history import decorations, pretty_value
from mirage.commands.cli.builtin.git.objects import abbrev_for
from mirage.commands.cli.builtin.git.repo import config_bool
from mirage.commands.cli.builtin.git.revparse import resolve_commit
from mirage.commands.cli.builtin.git.session import opened
from mirage.commands.cli.builtin.git.util import (  # yapf: disable
    check_operands, escaped, fatal, revision_arg)
from mirage.commands.cli.types import CLIDoors, CLIInvocation
from mirage.commands.spec.flag_view import FlagView
from mirage.io.stream import yield_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.shell.bytes import encode_text


@dataclass(frozen=True, slots=True)
class ShowFlags:
    """The commit presentation and shared diff options."""
    diff: DiffFlags
    pretty: LogFormat
    date: str = "default"


def parse_show_flags(fl: FlagView,
                     default_renames: bool = True,
                     quote_path_fully: bool = True) -> ShowFlags:
    """Read the raw show flag kwargs into a frozen struct.

    Args:
        fl (FlagView): spec-validated view over the raw flag kwargs.
        default_renames (bool): ``diff.renames``.
        quote_path_fully (bool): ``core.quotePath``.
    """
    spelled = pretty_value(fl)
    return ShowFlags(
        diff=parse_diff_flags(fl,
                              default_merge="dense-combined",
                              default_renames=default_renames,
                              quote_path_fully=quote_path_fully),
        date=fl.as_str("date") or "default",
        pretty=parse_pretty(spelled) if spelled is not None else MEDIUM,
    )


def _header(commit: Commit, flags: ShowFlags, width: int,
            decor: Decorations | None) -> bytes:
    """The commit header in the requested format.

    ``format:`` is a separator, so a single commit prints with no
    trailing newline at all; ``tformat:`` terminates the entry even
    when it renders empty, except that an empty template prints
    nothing, matching ``log --format=``. Pinned against git 2.37 and
    2.54.

    Args:
        commit (Commit): the commit being shown.
        flags (ShowFlags): the parsed invocation.
        width (int): abbreviated id width for this repository.
        decor (Decorations | None): ref labels when the format asked.
    """
    fmt = flags.pretty
    if fmt.kind == "oneline":
        return f"{oneline(commit, width)}\n".encode()
    if fmt.kind in ("format", "tformat"):
        rendered = render_template(fmt.template or "", commit, width, decor,
                                   flags.date)
        if fmt.kind == "tformat":
            return encode_text(f"{rendered}\n") if fmt.template else b""
        return encode_text(rendered)
    return ("\n".join(preset_block(commit, fmt.kind, width, flags.date)) +
            "\n").encode()


def _render(repo: BaseRepo, revision: str, flags: ShowFlags,
            want_decor: bool) -> bytes:
    """Resolve a revision and render its entry and diff, synchronously.

    Runs on a worker thread: resolving, walking the tree and reading
    blobs all fetch through the dispatcher, so this must not sit on the
    loop that answers those fetches.

    Args:
        repo (BaseRepo): repository to read.
        revision (str): the revision to show.
        flags (ShowFlags): the parsed invocation.
        want_decor (bool): whether the format renders %d/%D.
    """
    commit = resolve_commit(repo, revision)
    decor = decorations(repo) if want_decor else None
    header = _header(commit, flags, abbrev_for(repo), decor)
    bodies = commit_output(repo, commit, flags.diff)
    return join_output(commit, header, bodies, flags.pretty.kind,
                       abbrev_for(repo), flags.diff, flags.diff.summary
                       and not flags.diff.no_patch)


async def show(inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    """Show one commit: its log entry, then its diff against its parent.

    Args:
        inv (CLIInvocation[None]): the line's invocation record.
            git declares no config_model; the planes it reads
            (data through ``dispatch``, names through ``ns``) ride
            ``inv.doors``.
    """
    doors = inv.doors or CLIDoors()
    dispatch = doors.dispatch
    texts = inv.texts
    flags = inv.flags
    fl = FlagView(flags)
    try:
        if dispatch is None:
            raise NoWorkspaceError()
        check_operands(texts, marked=escaped(inv.argv))
        repo, location = await opened(fl, doors)
        parsed = parse_show_flags(
            fl, await renames_enabled(dispatch, location), await
            config_bool(dispatch, location, b"core", b"quotepath", True))
        rendered = await asyncio.to_thread(_render, repo, revision_arg(texts),
                                           parsed,
                                           needs_decorations(parsed.pretty))
    except GitError as exc:
        return fatal(exc)
    return yield_bytes(rendered), IOResult()


def _diff_tree(repo: BaseRepo, revision: str, flags: DiffFlags,
               no_commit_id: bool, recursive: bool) -> bytes:
    commit = resolve_commit(repo, revision)
    bodies = commit_output(repo, commit, flags, recursive, root=False)
    return b"".join(
        (b"" if no_commit_id else commit.id + b"\n") + body for body in bodies)


async def diff_tree(
        inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    doors = inv.doors or CLIDoors()
    try:
        if doors.dispatch is None:
            raise NoWorkspaceError()
        repo, location = await opened(fl, doors)
        fully = await config_bool(doors.dispatch, location, b"core",
                                  b"quotepath", True)
        out = await asyncio.to_thread(
            _diff_tree, repo, inv.texts[0] if inv.texts else "HEAD",
            parse_diff_flags(fl,
                             default_patch=False,
                             porcelain=False,
                             quote_path_fully=fully),
            fl.as_bool("no_commit_id"), fl.as_bool("r"))
        return out, IOResult()
    except GitError as exc:
        return fatal(exc)
